content/blog/2018-05-23-react-v-16-4.md

```diff
@@ -0,0 +1,128 @@
+---
+title: "React v16.4.0: Experimental Component Profiler and Bugfixes"
+author: [acdlite, bvaughn]
+---
+
+React 16.4 is a minor release that includes an experimental component profiler, as well as a crucial bugfix for `getDerivedStateFromProps`.
+
+## `React.unstable_Profiler`
+
+TODO (@bvaughn)
+
+## Bugfix for `getDerivedStateFromProps`
+
+`getDerivedStateFromProps` is now called when either props or state change. Previously, it would not fire for state-only changes (similar to `componentWillReceiveProps`). This was an oversight in the initial implementation that has now been corrected.
+
+### Why?
+
+`getDerivedStateFromProps` was designed as an alternative to `componentWillReceiveProps` that is compatible with async rendering. Because props in React are derived from state that lives in a parent component, APIs like `componentWillReceiveProps` that are modelled around "prop change events" don't really make sense, and can't be reliably implemented in React's upcoming asynchronous rendering mode.
+
+Another way to think about it is that `getDerivedStateFromProps(props, state)` accepts both props and state as input, so we need to re-calculate the derived state whenever either of the inputs change.
+
+### How Does This Affect Me?
+
+**Most likely, this doesn't require any changes to your components**. The rare cases where it does matter usually fall into one of two categories:
+
+### Avoid Side Effects in `getDerivedStateFromProps`
+
+Like the render method, `getDerivedStateFromProps` should be a pure function of props and state. You should move side-effectful code elsewhere. E.g. Flux dispatches typically belong inside the originating event handler; DOM mutations belong inside `componentDidMount` or `componentDidUpdate`.
+
+Side effects in `getDerivedStateFromProps` were already disallowed, but since it now fires more often than it used to, the recent change may expose previously undiscovered bugs.
+
+### Compare Incoming Props to Previous Props When Computing Controlled Values
+
+The following code will not work. It assumes `getDerivedStateFromProps` only fires on prop changes. But since it now fires for state changes, too, local updates to the controlled value will be ignored, since the props version always overrides it.
+
+```js
+static getDerivedStateFromProps(props, state) {
+  if (props.value !== state.controlledValue) {
```

Comment:
It might be worth mentioning that this pattern was flawed to begin with- even setting the recent gDSFP change aside- because it would override `state` even when unrelated `props` changed.

Reply:
I don't know if that would be more or less confusing. It might distract from the thing you're trying to emphasize. I don't know.